File: progressSheet/valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>> &board) {
        int masks[9] = {0, 0, 0, 0, 0, 0, 0, 0, 0};
        int bmasks[9] = {0, 0, 0, 0, 0, 0, 0, 0, 0};

        for (int i = 0; i < 9; ++i) {
            int mask = 0;
            for (int j = 0; j < 9; ++j) {
                if (board[i][j] == '.')
                    continue;

                int ind = 1 << (board[i][j] - '1');
                if (mask & ind)
                    return false;

                if (masks[j] & ind)
                    return false;

                int r = i / 3, c = j / 3;
                int b = r * 3 + c;

                if (bmasks[b] & ind)
                    return false;

                mask |= ind;
                masks[j] |= ind;
                bmasks[b] |= ind;
            }
        }

        return true;
    }
};

File: progressSheet/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        def isValid(row: int, col: int) -> bool:
            boxRow = 3 * (row // 3)
            boxCol = 3 * (col // 3)

            for index in range(9):
                if index == col or board[row][index] == '.':
                    continue
                
                if board[row][index] == board[row][col]:
                    return False
            
            for index in range(9):
                if index == row or board[index][col] == '.':
                    continue
                
                if board[index][col] == board[row][col]:
                    return False
            
            for x in range(3):
                for y in range(3):
                    r = boxRow + x
                    c = boxCol + y

                    if r == row and c == col or board[r][c] == '.':
                        continue
                    
                    if board[r][c] == board[row][col]:
                        return False
            
            return True
                

        for row in range(9):
            for col in range(9):
                if not isValid(row, col):
                    return False
        
        return True